### PR TITLE
feat(Kinopub): add presence

### PR DESCRIPTION
**Pull request by shiseni**
## Description 
Add presence for Kinopub

Settings:
1. Private mode.
2. Show remaining listening/viewing time.
3. Showing movie/serial covers.
4. Show buttons.

## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

image
image
image

</details>

### CODE REVIEW

File: websites/K/Kinopub/metadata.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
+{
+  "$schema": "https://schemas.premid.app/metadata/1.7",
+  "service": "Kinopub",
+  "author": {
+    "id": "257151906038677504",
+    "name": "SHISEN#8023"
+  },
+  "description": {
+    "en": "Website for online watching movies, series, sports broadcasts and concerts",
```
Comment (DarkVillager) [inline websites/K/Kinopub/metadata.json:9]: I'd also mention the podcast part (cuz the official thing mentions it)
```suggestion
    "en": "Website for listening to podcasts and watching movies, series, sports broadcasts and concerts all online",
```

File: websites/K/Kinopub/metadata.json [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	"iframe": true,
+	"settings": [
+		{
+			"id": "privacy",
+			"title": "Приватный режим",
+			"icon": "fad fa-user-secret",
+			"value": false
+		},
+		{
+			"id": "time",
```
Comment (DarkVillager) [inline websites/K/Kinopub/metadata.json:33]:
```suggestion
			"id": "timestamps",
```

File: websites/K/Kinopub/presence.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
+const presence = new Presence({
+	clientId: "1054755173198737458",
+});
+
+enum Assets {
+	Play = "https://i.imgur.com/6s4WyWY.png",
+	Pause = "https://i.imgur.com/PrYtpQb.png",
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:7]: Change it to 512x512px
```suggestion
	Pause = "https://i.imgur.com/4iyMINk.png",
```

File: websites/K/Kinopub/presence.ts [hunk 4]
```diff
@@ -0,0 +1,254 @@
+const presence = new Presence({
+	clientId: "1054755173198737458",
+});
+
+enum Assets {
+	Play = "https://i.imgur.com/6s4WyWY.png",
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:6]: Change it to 512x512px
```suggestion
	Play = "https://i.imgur.com/OLaz6JN.png",
```

File: websites/K/Kinopub/presence.ts [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+	Pause = "https://i.imgur.com/PrYtpQb.png",
+	Read = "https://i.imgur.com/wPUmqu5.png",
+	View = "https://i.imgur.com/hxvvGUi.png",
+	Search = "https://i.imgur.com/wYVlwJX.png",
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:10]:
```suggestion
	Search = "https://i.imgur.com/oGQtnIY.png"",
```

File: websites/K/Kinopub/presence.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
+const presence = new Presence({
+	clientId: "1054755173198737458",
+});
+
+enum Assets {
+	Play = "https://i.imgur.com/6s4WyWY.png",
+	Pause = "https://i.imgur.com/PrYtpQb.png",
+	Read = "https://i.imgur.com/wPUmqu5.png",
+	View = "https://i.imgur.com/hxvvGUi.png",
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:9]: Change it to 512x512px
```suggestion
	View = "https://i.imgur.com/KN6uwcg.png",
```

File: websites/K/Kinopub/presence.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
+const presence = new Presence({
+	clientId: "1054755173198737458",
+});
+
+enum Assets {
+	Play = "https://i.imgur.com/6s4WyWY.png",
+	Pause = "https://i.imgur.com/PrYtpQb.png",
+	Read = "https://i.imgur.com/wPUmqu5.png",
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:8]: Change it to 512x512px & use the correct name like in .resources
```suggestion
	Reading = "https://i.imgur.com/MG2O4Ks.png",
```

File: websites/K/Kinopub/presence.ts [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+					} else {
+						[presenceData.startTimestamp, presenceData.endTimestamp] =
+							presence.getTimestamps(video.currentTime, video.duration);
+					}
+				}
+			}
+			break;
+	}
+
+	if (privacy) delete presenceData.state;
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:251]: I'd remove this and just use the following before the switch statement
```ts
	if (privacy) {
		presenceData.details = strings.browse;
		presence.setActivity(presenceData);
		return;
	}
```

File: websites/K/Kinopub/presence.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+	if (privacy || !buttons) delete presenceData.buttons;
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:252]: If the comment above so is done replace it with this:
```suggestion
	if (presenceData.buttons && !buttons) delete presenceData.buttons;
```

File: websites/K/Kinopub/presence.ts [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+	presence.setActivity(presenceData);
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:253]: This suggestion is optional, but this makes it just show the logo if it errors or sum.
```suggestion
	if (presenceData.details) presence.setActivity(presenceData);
	else presence.setActivity();
```

**Review by DarkVillager — CHANGES_REQUESTED** [review]

File: websites/K/Kinopub/presence.ts [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+	(data: { duration: number; currentTime: number; paused: boolean }) => {
+		video = data;
+	}
+);
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			details: "Где-то на сайте",
+			largeImageKey: "https://i.imgur.com/HAvvvj7.png",
+			smallImageText: "Kinopub",
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:60]: Let's not do that.
```suggestion
```

File: websites/K/Kinopub/presence.ts [hunk 12]
```diff
@@ -0,0 +1,254 @@
+const presence = new Presence({
+	clientId: "1054755173198737458",
+});
+
+enum Assets {
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:5]:
```suggestion
enum Assets {
	Logo = "https://i.imgur.com/HAvvvj7.png",
```

File: websites/K/Kinopub/presence.ts [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+	"iFrameData",
+	(data: { duration: number; currentTime: number; paused: boolean }) => {
+		video = data;
+	}
+);
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			details: "Где-то на сайте",
+			largeImageKey: "https://i.imgur.com/HAvvvj7.png",
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:59]:
```suggestion
			largeImageKey: Assets.Logo,
```

File: websites/K/Kinopub/presence.ts [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+			smallImageText: "Kinopub",
+		},
+		[privacy, time, buttons, cover] = await Promise.all([
+			presence.getSetting<boolean>("privacy"),
+			presence.getSetting<boolean>("time"),
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:64]:
```suggestion
			presence.getSetting<boolean>("timestamps"),
```

File: websites/K/Kinopub/presence.ts [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+	}
+);
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			details: "Где-то на сайте",
+			largeImageKey: "https://i.imgur.com/HAvvvj7.png",
+			smallImageText: "Kinopub",
+		},
+		[privacy, time, buttons, cover] = await Promise.all([
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:62]:
```suggestion
		[privacy, timestamps, buttons, cover] = await Promise.all([
```

File: websites/K/Kinopub/presence.ts [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+						)?.childNodes[0]?.textContent;
+					presenceData.state = serieName;
+					presenceData.smallImageKey = video.paused
+						? Assets.Pause
+						: Assets.Play;
+					presenceData.smallImageText = video.paused
+						? Strings.Pause
+						: Strings.Play;
+
+					if (video.paused || !time) {
```
Comment (DarkVillager) [inline websites/K/Kinopub/presence.ts:239]:
```suggestion
					if (video.paused || !timestamps) {
```

**Review by DarkVillager — CHANGES_REQUESTED** [review]

File: websites/K/Kinopub/metadata.json (hunk 1 above)
Comment (shiseni) [reply websites/K/Kinopub/metadata.json:9]: > I'd also mention the podcast part cuz the official thing mentions it. (Also changed the Russian to match the English one.

Website does not have a listening to podcasts feature.

File: websites/K/Kinopub/presence.ts (hunk 3 above)
Comment (shiseni) [reply websites/K/Kinopub/presence.ts:8]: > Change it to 512x512px

Same image.

File: websites/K/Kinopub/presence.ts (hunk 8 above)
Comment (shiseni) [reply websites/K/Kinopub/presence.ts:251]: > I'd remove this and just use the following before the switch statement
> 
> ```ts
> 	if (privacy) {
> 		presenceData.details = strings.browse;
> 		presence.setActivity(presenceData);
> 		return;
> 	}
> ```

Not suitable for my case.

File: websites/K/Kinopub/presence.ts (hunk 10 above)
Comment (shiseni) [reply websites/K/Kinopub/presence.ts:253]: > This suggestion is optional, but this makes it just show the logo if it errors or sum.

By default, if the path is not found, "Somewhere on the site" is displayed.

File: websites/K/Kinopub/metadata.json (hunk 1 above)
Comment (DarkVillager) [reply websites/K/Kinopub/metadata.json:9]: ok

File: websites/K/Kinopub/presence.ts (hunk 3 above)
Comment (DarkVillager) [reply websites/K/Kinopub/presence.ts:8]: Change the size.

**Review by EncryptedDev — APPROVED** [review]

**Review by Bas950 — APPROVED** [review]

**Pull request merged**